### PR TITLE
Add @@species accessor to Array and Promise builtin objects

### PR DESCRIPTION
**Pull request by DanielBallaSZTE**
Fully support @@species and SpeciesConstructor in Array and Promise builtins.
Also added partial support to TypedArrays, but a rework is needed in %TypedArray%.prototype functions' typedarray constructor, which is out of this patch's scope.

JerryScript-DCO-1.0-Signed-off-by: Daniel Balla dballa@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-promise.c [hunk 1]
```diff
@@ -574,15 +580,22 @@ ecma_builtin_promise_race_or_all (ecma_value_t this_arg, /**< 'this' argument */
...
 
-  if (builtin_id != ECMA_BUILTIN_ID_PROMISE)
+  if (ECMA_IS_VALUE_ERROR (species_symbol))
   {
-    return ecma_raise_type_error (ECMA_ERR_MSG ("'this' is not the Promise constructor."));
+    return species_symbol;
   }
 
-  ecma_value_t capability = ecma_promise_new_capability ();
+  ecma_value_t species = ecma_op_species_constructor (ecma_get_object_from_value (this_arg),
```
Comment (dbatyai) [inline jerry-core/ecma/builtin-objects/ecma-builtin-promise.c]: I don't think we need to call `SpeciesConstructor` here.

File: jerry-core/ecma/operations/ecma-array-object.c [hunk 2]
```diff
@@ -663,27 +663,69 @@ ecma_op_create_array_object (const ecma_value_t *arguments_list_p, /**< list of
...
-                                                                                 *   arguments_list_len is 1
-                                                                                 *   and single argument is Number,
-                                                                                 *   then treat the single argument
-                                                                                 *   as new Array's length rather
-                                                                                 *   than as single item of the
-                                                                                 *   Array */
-                                            ecma_object_t *object_p) /**< The object from whom the new array object
-                                                                      *   is being created */
+ecma_op_array_species_create (ecma_object_t *original_array_p, /**< The object from whom the new array object
+                                                        *   is being created */
```
Comment (dbatyai) [inline jerry-core/ecma/operations/ecma-array-object.c]: This commend is misaligned.

File: jerry-core/ecma/operations/ecma-array-object.c [hunk 3]
```diff
@@ -663,27 +663,69 @@ ecma_op_create_array_object (const ecma_value_t *arguments_list_p, /**< list of
...
+    if (ecma_is_constructor (constructor)
+        && ecma_get_object_from_value (constructor) == ecma_builtin_get (ECMA_BUILTIN_ID_ARRAY))
+    {
+      ecma_free_value (constructor);
+      constructor = ECMA_VALUE_UNDEFINED;
+    }
+    else if (ecma_is_value_object (constructor))
+    {
+      ecma_object_t *ctor_object_p = ecma_get_object_from_value (constructor);
+      ecma_value_t species = ecma_op_object_get_by_symbol_id (ctor_object_p, LIT_MAGIC_STRING_SPECIES);
```
Comment (dbatyai) [inline jerry-core/ecma/operations/ecma-array-object.c]: Could we assign directly to `constructor`?

File: jerry-core/ecma/operations/ecma-array-object.c [hunk 4]
```diff
@@ -663,27 +663,69 @@ ecma_op_create_array_object (const ecma_value_t *arguments_list_p, /**< list of
...
+    return ecma_make_object_value (ecma_op_new_fast_array_object (length));
+  }
+
+  if (!ecma_is_constructor (constructor))
+  {
+    ecma_free_value (constructor);
+    return ecma_raise_type_error (ECMA_ERR_MSG ("TODO"));
+  }
+
+  ecma_value_t len_val = ecma_make_uint32_value (length);
```
Comment (dbatyai) [inline jerry-core/ecma/operations/ecma-array-object.c]: The standard specifies extra steps for negative zero length, do we have that handled somewhere?

Comment (DanielBallaSZTE) [reply jerry-core/ecma/operations/ecma-array-object.c]: `length` is an `ecma_length_t` which is an `uint32_t`, therefore we can only pass +0

File: jerry-core/ecma/operations/ecma-objects.c [hunk 5]
```diff
@@ -2614,6 +2614,62 @@ ecma_op_is_regexp (ecma_value_t arg) /**< argument */
...
+ * SpeciesConstructor operation
+ * See also:
+ *          ECMA-262 v6, 7.3.20;
+ *
+ * @return ecma_value
+ *         returned value must be freed with ecma_free_value
+ */
+ecma_value_t
+ecma_op_species_constructor (ecma_object_t *this_value,
+                             ecma_object_t *default_constructor)
```
[the remark this answers is not in the document]
Comment (DanielBallaSZTE) [reply jerry-core/ecma/operations/ecma-objects.c]: Good idea

File: jerry-core/ecma/operations/ecma-objects.c [hunk 6]
```diff
@@ -2614,6 +2614,63 @@ ecma_op_is_regexp (ecma_value_t arg) /**< argument */
 
   return ecma_make_boolean_value (ecma_object_is_regexp_object (arg));
 } /* ecma_op_is_regexp */
+
+/**
+ *
```
Comment (dbatyai) [inline jerry-core/ecma/operations/ecma-objects.c]: Empty line.

File: jerry-core/ecma/operations/ecma-promise-object.c [hunk 7]
```diff
@@ -573,6 +573,68 @@ ecma_op_create_promise_object (ecma_value_t executor, /**< the executor function
   return ecma_make_object_value (object_p);
 } /* ecma_op_create_promise_object */
 
+/**
+ * 25.4.1.5.1 GetCapabilitiesExecutor Functions
```
Comment (dbatyai) [inline jerry-core/ecma/operations/ecma-promise-object.c]: Could you improve this description a bit?

File: jerry-core/ecma/operations/ecma-objects.c [hunk 8]
```diff
@@ -2614,6 +2614,63 @@ ecma_op_is_regexp (ecma_value_t arg) /**< argument */
...
+ * SpeciesConstructor operation
+ * See also:
+ *          ECMA-262 v6, 7.3.20;
+ *
+ * @return ecma_value
+ *         returned value must be freed with ecma_free_value
+ */
+ecma_value_t
+ecma_op_species_constructor (ecma_object_t *this_value,
+                             ecma_builtin_id_t default_constructor_id)
```
Comment (dbatyai) [inline jerry-core/ecma/operations/ecma-objects.c]: Missing comments for arguments.

**Pull request merged**